runtime/sema/checker.go

```diff
@@ -1810,181 +1809,210 @@ func (checker *Checker) checkDeclarationAccessModifier(
 				},
 			)
 		}
-	} else {
+		return
+	}
 
-		isTypeDeclaration := declarationKind.IsTypeDeclaration()
-
-		switch access := access.(type) {
-		case PrimitiveAccess:
-			switch ast.PrimitiveAccess(access) {
-			case ast.AccessPublicSettable:
-				// Public settable access for a constant is not sensible
-				// and type declarations must be public for now
-
-				if isConstant || isTypeDeclaration {
-					var explanation string
-					switch {
-					case isConstant:
-						explanation = "constants can never be set"
-					case isTypeDeclaration:
-						explanation = invalidTypeDeclarationAccessModifierExplanation
-					}
+	switch access := access.(type) {
+	case PrimitiveAccess:
+		checker.checkPrimitiveAccess(access, isConstant, declarationKind, startPos)
+	case EntitlementMapAccess:
+		checker.checkEntitlementMapAccess(access, declarationKind, declarationType, containerKind, startPos)
+	case EntitlementSetAccess:
+		checker.checkEntitlementSetAccess(declarationType, containerKind, startPos)
+	}
```

Comment:
Refactored this switch case, because the method was getting pretty large (moved to functions as-is)
Check the PR diff with the "Hide whitespace" option to see the actual code changes.